cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java

```diff
@@ -608,7 +608,7 @@ private SearchResults searchByCustomIndex(String namespaceId, Set<EntityTypeSimp
     String indexColumn = getIndexColumn(sortInfo.getSortBy(), sortInfo.getSortOrder());
     // we want to return the first chunk of 'limit' elements after offset
     // in addition, we want to pre-fetch 'numCursors' chunks of size 'limit'
-    int fetchSize = offset + ((numCursors + 1) * limit);
+    int fetchSize = (int) Math.min(offset + ((numCursors + 1) * (long) limit), Integer.MAX_VALUE);
```

Comment:
ok. Is this easier to read then?
```
if (Integer.MAX_VALUE == limit) {
  fetchSize = limit;
} else {
  fetchSize = offset + ((numCursors + 1) * limit);
}
```

It's ok to keep it the way it is, but please add a comment explaining why this is done.